File: ConceptOfTheDay/Python/src/kadane_algorithm.py
"""
@author - Anirudh Sharma
"""


def maximum_sum(a):
    if a is None or len(a) == 0:
        return 0
    global_maximum = a[0]
    local_maximum = a[0]
    for i in range(1, len(a)):
        local_maximum = max(a[i], local_maximum + a[i])
        if global_maximum < local_maximum:
            global_maximum = local_maximum
    return global_maximum


a = [-2, -3, 4, -1, -2, 1, 5, -3]
print(maximum_sum(a))

a = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
print(maximum_sum(a))

a = [0, 1, 2, -2, 3, 2]
print(maximum_sum(a))

a = [-2, 2, 5, -11, 6]
print(maximum_sum(a))

File: ConceptOfTheDay/JavaScript/src/kadane_algorithm.js
function maximum_sum(a) {
    if (!a || a.length == 0) {
        return 0;
    }
    let localMaximum = a[0];
    let globalMaximum = a[0];
    for(let i = 0; i < a.length; i++) {
        localMaximum = Math.max(a[i], localMaximum + a[i]);
        if(globalMaximum < localMaximum) {
            globalMaximum = localMaximum;
        }
    }
    return globalMaximum;
}

a = [-2, -3, 4, -1, -2, 1, 5, -3];
console.log(maximum_sum(a));

a = [-2, 1, -3, 4, -1, 2, 1, -5, 4];
console.log(maximum_sum(a));

a = [0, 1, 2, -2, 3, 2];
console.log(maximum_sum(a));

a = [-2, 2, 5, -11, 6];
console.log(maximum_sum(a));